### ISSUE TITLE
Sum of sigmas when computing grad to estimate normals

### ISSUE THREAD

**Issue by mra-h** [justin871030/refnerf-pl#1]
Hello,

Thanks for the implementation!

I have a question regarding on how to compute normals using `torch.autograd.grad`. [Here](https://github.com/justin871030/refnerf-pl/blob/3d15a6d143eb1ff2a9b02ac4bf5b9101733d6219/internal/models.py#L609) you do sigmas.sum() as output tensor. But why is the sum necessary? 

Is it the same to compute the gradient as 
`torch.autograd.grad(raw_density, means, grad_outputs=torch.ones_like(raw_density), retain_graph=True)`. If not, you know the reason why?

Thanks!

**Comment by minfenli** [comment]
Happy to answer your question! 

The reason to use `sigmas.sum()` is to backpropagate the gradients of `sigmas` to their corresponding `means`(coordinates) with one function call. It's equal to compute each gradient `sigma` with respect to its input `mean`.

`.sum()` is like connecting the gradient graph to a single node without influence gradient values.

I'm not sure whether `torch.autograd.grad(raw_density, means, grad_outputs=torch.ones_like(raw_density), retain_graph=True)` will give the same result.

**Comment by mra-h** [comment]
Thanks for the reply. I have tested it and both approaches yield the same result. I think as you said with `.sum()` you create a single output node in the graph. The gradient of a `.sum()` wrt to the argument inside the sum is 1, i.e., it is the same as not computing the sum and provide  `grad_outputs=torch.ones_like(raw_density)`
